### PR TITLE
2nd Refactoring

### PR DESCRIPTION
**Pull request by vince62s**
this includes:
- torchtext 0.3
- average attention network
- various refactoring follow-ups

@srush  ready for review

### CODE REVIEW

File: onmt/__init__.py [hunk 1]
```diff
@@ -1,8 +1,20 @@
...
 import onmt.utils.optimizers
 onmt.utils.optimizers.Optim = onmt.utils.optimizers.Optimizer
 sys.modules["onmt.Optim"] = onmt.utils.optimizers
 
-__all__ = ["Trainer"]
+# For Flake
+__all__ = [onmt.inputters, onmt.encoders, onmt.decoders, onmt.models,
+           onmt.utils, onmt.modules, "Trainer"]
+
+__version__ = "0.4.0"
```
Comment (srush) [inline onmt/__init__.py]: Is this the current OpenNMT version #?

File: onmt/decoders/transformer.py [hunk 2]
```diff
@@ -8,8 +8,8 @@
 
 import onmt
 from onmt.decoders.decoder import DecoderState
-from onmt.utils.misc import aeq
-from onmt.utils.transformer_util import PositionwiseFeedForward
+# from onmt.utils.misc import aeq
```
Comment (srush) [inline onmt/decoders/transformer.py]: delete

File: onmt/decoders/transformer.py [hunk 3]
```diff
@@ -45,22 +53,36 @@ def __init__(self, size, dropout,
...
+        Returns:
+            output: [ batch_size, 1, model_dim ]
+            attn: [ batch_size, 1, src_len ]
+            all_input: [ batch_size, current_step, model_dim ]
 
-        src_batch, _, s_len = src_pad_mask.size()
-        tgt_batch, _, _ = tgt_pad_mask.size()
+
+        """
+        # Args Checks - commented out for performance issue
```
Comment (srush) [inline onmt/decoders/transformer.py]: Either add back or remove.

File: onmt/decoders/transformer.py [hunk 4]
```diff
@@ -45,22 +53,36 @@ def __init__(self, size, dropout,
...
-        # Args Checks
-        input_batch, input_len, _ = inputs.size()
-        if previous_input is not None:
-            pi_batch, _, _ = previous_input.size()
-            aeq(pi_batch, input_batch)
-        contxt_batch, contxt_len, _ = memory_bank.size()
-        aeq(input_batch, contxt_batch)
+                previous_input=None, layer_cache=None, step=None):
+        """
+        Args:
```
Comment (srush) [inline onmt/decoders/transformer.py]: Please make these comments in the same style as the rest of the code, as it is autodoc'd. Please see other places in the code. Particularly the return format.

File: onmt/decoders/transformer.py [hunk 5]
```diff
@@ -81,20 +109,28 @@ def forward(self, inputs, memory_bank, src_pad_mask, tgt_pad_mask,
...
         # CHECKS
-        output_batch, output_len, _ = output.size()
-        aeq(input_len, output_len)
-        aeq(contxt_batch, output_batch)
-
-        n_batch_, t_len_, s_len_ = attn.size()
-        aeq(input_batch, n_batch_)
-        aeq(contxt_len, s_len_)
-        aeq(input_len, t_len_)
+        # output_batch, output_len, _ = output.size()
```
Comment (srush) [inline onmt/decoders/transformer.py]: Fix or remove.

File: onmt/decoders/transformer.py [hunk 6]
```diff
@@ -81,20 +109,28 @@ def forward(self, inputs, memory_bank, src_pad_mask, tgt_pad_mask,
...
+        # aeq(input_batch, n_batch_)
+        # aeq(contxt_len, s_len_)
+        # aeq(input_len, t_len_)
         # END CHECKS
 
         return output, attn, all_input
 
     def _get_attn_subsequent_mask(self, size):
-        ''' Get an attention mask to avoid using the subsequent info.'''
+        """
```
Comment (srush) [inline onmt/decoders/transformer.py]: Fix comment format.

File: onmt/decoders/transformer.py [hunk 7]
```diff
@@ -153,27 +189,36 @@ def __init__(self, num_layers, hidden_size, attn_type,
             self._copy = True
         self.layer_norm = onmt.modules.LayerNorm(hidden_size)
 
-    def forward(self, tgt, memory_bank, state, memory_lengths=None):
+    def _init_cache(self, memory_bank, memory_lengths=None):
```
Comment (srush) [inline onmt/decoders/transformer.py]: Cool, I like this.

File: onmt/decoders/transformer.py [hunk 8]
```diff
@@ -153,27 +189,36 @@ def __init__(self, num_layers, hidden_size, attn_type,
...
+                step=None, cache=None):
         """
         See :obj:`onmt.modules.RNNDecoderBase.forward()`
         """
         # CHECKS
-        assert isinstance(state, TransformerDecoderState)
-        tgt_len, tgt_batch, _ = tgt.size()
-        memory_len, memory_batch, _ = memory_bank.size()
-        aeq(tgt_batch, memory_batch)
+        # assert isinstance(state, TransformerDecoderState)
```
Comment (srush) [inline onmt/decoders/transformer.py]: Fix or remove.

File: onmt/encoders/transformer.py [hunk 9]
```diff
@@ -90,14 +99,15 @@ def forward(self, src, lengths=None):
 
         out = emb.transpose(0, 1).contiguous()
         words = src[:, :, 0].transpose(0, 1)
-        # CHECKS
-        out_batch, out_len, _ = out.size()
-        w_batch, w_len = words.size()
-        aeq(out_batch, w_batch)
-        aeq(out_len, w_len)
+        # CHECKS - commented for performance issue
```
Comment (srush) [inline onmt/encoders/transformer.py]: Fix or remove.

File: onmt/encoders/transformer.py [hunk 10]
```diff
@@ -61,14 +69,15 @@ class TransformerEncoder(EncoderBase):
...
-       num_layers (int): number of encoder layers
-       hidden_size (int): number of hidden units
-       dropout (float): dropout parameters
-       embeddings (:obj:`onmt.modules.Embeddings`):
+        num_layers (int): number of encoder layers
+        hidden_size (int): number of hidden units
+        dropout (float): dropout parameters
+        embeddings (:obj:`onmt.modules.Embeddings`):
           embeddings to use, should have positional encodings
+    Returns:
```
Comment (srush) [inline onmt/encoders/transformer.py]: Fix comment style

File: onmt/encoders/transformer.py [hunk 11]
```diff
@@ -38,6 +36,16 @@ def __init__(self, size, dropout,
...
     def forward(self, inputs, mask):
+        """
+        Transformer Encoder Layer definition.
+
+        Args:
+            inputs: [ batch_size, src_len, model_dim ]
+            mask: [ batch_size, src_len, src_len ]
+
+        Returns:
+            outputs: [ batch_size, src_len, model_dim ]
```
Comment (srush) [inline onmt/encoders/transformer.py]: Fix comment style.

File: onmt/inputters/inputter.py [hunk 12]
```diff
@@ -75,7 +75,7 @@ def save_fields_to_vocab(fields):
     vocab = []
     for k, f in fields.items():
         if f is not None and 'vocab' in f.__dict__:
-            f.vocab.stoi = dict(f.vocab.stoi)
+            f.vocab.stoi = f.vocab.stoi
```
Comment (srush) [inline onmt/inputters/inputter.py]: Does this line do anything? We needed to make it a dict before to save. Remove or check.

File: onmt/models/model.py [hunk 13]
```diff
@@ -1,6 +1,4 @@
 """ Onmt NMT Model base class definition """
```
Comment (srush) [inline onmt/models/model.py]: We need this for 2.7.

File: onmt/modules/average_attn.py [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+        else:
+            mask = mask_or_step
+            return torch.matmul(mask, inputs)
+
+    def forward(self, inputs, mask=None, layer_cache=None, step=None):
+        """
+        Args:
+          inputs: [ batch_size, input_len, model_dim ]
+
+        Returns:
```
Comment (srush) [inline onmt/modules/average_attn.py]: Fix comment.

File: onmt/modules/copy_generator.py [hunk 15]
```diff
@@ -65,6 +64,8 @@ def __init__(self, input_size, tgt_dict):
         self.linear = nn.Linear(input_size, len(tgt_dict))
         self.linear_copy = nn.Linear(input_size, 1)
         self.tgt_dict = tgt_dict
+        self.softmax = nn.Softmax()
```
Comment (srush) [inline onmt/modules/copy_generator.py]: This is fine, but you don't actually need to do this (they are parameter-less). Do you like the style better?

File: onmt/modules/embeddings.py [hunk 16]
```diff
@@ -182,14 +184,14 @@ def forward(self, source):
         Return:
             `FloatTensor`: word embeddings `[len x batch x embedding_size]`
         """
-        in_length, in_batch, nfeat = source.size()
-        aeq(nfeat, len(self.emb_luts))
+        # in_length, in_batch, nfeat = source.size()
```
Comment (srush) [inline onmt/modules/embeddings.py]: Drop or fix.

File: onmt/modules/embeddings.py [hunk 17]
```diff
@@ -40,6 +40,8 @@ def forward(self, emb):
         emb = self.dropout(emb)
         return emb
 
+# used by model_builder.py and subsequently by all encoders / decoders
```
Comment (srush) [inline onmt/modules/embeddings.py]: Drop.

File: onmt/modules/global_attention.py [hunk 18]
```diff
@@ -4,6 +4,11 @@
 
 from onmt.utils.misc import aeq, sequence_mask
 
+# This class is mainly used by decoder.py for RNNs but also
```
Comment (srush) [inline onmt/modules/global_attention.py]: Make a docstring.

File: onmt/modules/multi_headed_attn.py [hunk 19]
```diff
@@ -87,20 +87,20 @@ def forward(self, key, value, query, mask=None):
...
-        batch_, q_len, d_ = query.size()
-        aeq(batch, batch_)
-        aeq(d, d_)
-        aeq(self.model_dim % 8, 0)
-        if mask is not None:
-            batch_, q_len_, k_len_ = mask.size()
-            aeq(batch_, batch)
-            aeq(k_len_, k_len)
-            aeq(q_len_ == q_len)
+        # batch, k_len, d = key.size()
```
Comment (srush) [inline onmt/modules/multi_headed_attn.py]: Drop or fix.

File: onmt/modules/position_ffn.py [hunk 20]
```diff
@@ -0,0 +1,44 @@
+"""
```
Comment (srush) [inline onmt/modules/position_ffn.py]: Cool thanks!

File: onmt/translate/translator.py [hunk 21]
```diff
@@ -142,9 +145,29 @@ def translate(self,
...
+            src_data_iter (iterator): an interator generating source data
+                e.g. it may be a list or an openned file
+            tgt_path (str): filepath of target data
+            tgt_data_iter (iterator): an interator generating target data
+            src_dir (str): source directory path
+                (used for Audio and Image datasets)
+            batch_size (int): size of examples per mini-batch
+            attn_debug (bool): enables the attention logging
+
+        Returns:
```
Comment (srush) [inline onmt/translate/translator.py]: fix return.

File: onmt/translate/translator.py [hunk 22]
```diff
@@ -266,7 +297,10 @@ def translate_batch(self, batch, data):
         Todo:
            Shouldn't need the original dataset.
         """
+        with torch.no_grad():
```
Comment (srush) [inline onmt/translate/translator.py]: nice.

File: onmt/translate/translator.py [hunk 23]
```diff
@@ -345,9 +386,16 @@ def unbottle(m):
             inp = inp.unsqueeze(2)
 
             # Run one step.
-            dec_out, dec_states, attn = self.model.decoder(
-                inp, memory_bank, dec_states, memory_lengths=memory_lengths)
+            if self.self_attn_type == "average":
```
Comment (srush) [inline onmt/translate/translator.py]: Would prefer this not to be in `translate.py` which is a key pathway. Any way you can have this work automatically?

File: onmt/utils/optimizers.py [hunk 24]
```diff
@@ -1,11 +1,10 @@
 #!/usr/bin/env python
 """ Optimizers class """
-from __future__ import print_function
```
Comment (srush) [inline onmt/utils/optimizers.py]: Need for 2.7 support.

File: requirements.txt [hunk 25]
```diff
@@ -1,5 +1,5 @@
 six
 tqdm
 torch>=0.4.0
-torchtext>=0.2.3
+git+https://github.com/pytorch/text
```
Comment (srush) [inline requirements.txt]: We need a version here or we can host our own. Torchtext will break if we point to their github.

File: setup.py [hunk 26]
```diff
@@ -4,7 +4,8 @@
 
 setup(name='OpenNMT-py',
       description='A python implementation of OpenNMT',
-      version='0.1',
+      version='0.4',
```
Comment (srush) [inline setup.py]: Why not 0.2?

**Comment by vince62s** [comment]
Thanks for the review Sasha.
Version: it was just out of my head to be in line with pytorch 0.4
If you say 0.2 fine to me. But in a way, we should head up to 1.0 quite soon.
AEQ: it does change the speed. dont' have numbers right now but when tuning I found it has an impact.

doc stuff: I am not 100% familiar on how it works, will do my best.

average attn: @francoishernandez can you have a look ? thanks.

torchtext: can we specify a version when pointing to their github ? right now they have not released 0.3

**Comment by srush** [comment]
* Yeah, let's say 0.2. Maybe we can aim to his 1.0 with pytorch, they might have more breaking changes. 
* AEQ: How are you profiling? It's hard to believe that a size check could really change total speed. It might however change the evaluation order. I am fine with removing these from the code. Maybe they can become tests. 
* Docs: Oh sorry, I can be more clear. The format has to be exactly like this one: 
https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/modules/global_attention.py#L133 

You can't have multiple returns argumetns, but you can return a tuple.

* I can help fix the avg attn thing in revision, if that's hard. The rule is: when possible translate should be model agnostic. 

* What if we just fork the current version to opennmt/torchtext?  I'm just worried using their github, as in the past they have updated their code with breaking changes.

File: onmt/inputters/inputter.py (hunk 12 above)
Comment (vince62s) [reply onmt/inputters/inputter.py]: @pltrdy any comment on this one ?
I know that we had crashing issue while the dict() was there.
to me it's unclear.

**Comment by rbommasanimozilla** [comment]
Quick question: What does the above PR/commits effect in terms of progress in MultiGPU/what else is needed to get MultiGPU to work and is there any easy way to help if more work is needed?

**Comment by srush** [comment]
This is laying the ground work for Pr3 which will introduce multigpu into the main codebase.

**Pull request merged**